contributors/design-proposals/federated-deployment.md

```diff
@@ -0,0 +1,127 @@
+    + Specifying min/max/weight for each of the clusters
+    + Be able to rebalance replicas once a cluster is added, lost or if there is no capacity.
+  + [R5] Each of the underlying clusters should be able to work independently if the federation control plane is down,
+  or there are some cross-cluster network issues. The clusters, when federation controll plane is down, shoule be as easily updatable/managable as if federation were present. For high availability, federation controll plane should be considered an "addon" that makes some extra features possible, not a component that brings or replaces the core K8S functionality. 
+  This basically means that Federated Deployment should create fully-functional deployments in the clusters. 
+
+## Design
+
+Regular Deployments don’t maintain the replicas by themselves. They let ReplicaSets keep track of them. Each of the deployments create a separate ReplicaSet for each of the pod template versions they have. To do a rolling upgrade the deployment slowly adds replicas to the replica set with the new pod template and takes them from the old one. The names of ReplicaSets are in form of deployment_name + “-” + podspec_hash. So it is relatively easy to tell which replicaset is responsible for pods with the given spec (if you are a machine of course). 
+
+<img src="federated-deployment-1.png" width="400"/>
+
+When adding the federation layer over the deployment the things get more complicated. Federated deployment needs to control the regular deployment. That’s understandable and in line with other controllers. Now what about ReplicaSets? Requirements [R1] [R2] [R3] imply that there should be ReplicaSets at the federation level. 
+
+<img src="federated-deployment-2.png" width="500"/>
+
+However Federated ReplicaSets should not control the ReplicaSets in the underlying clusters because they are already controlled by the local Deployments and having two controllers would lead to race condition issues. Two controllers independently trying to bring the current state to the same desired state will eventually do it, but on the way multiple back and forth steps can be executed. The user looking at this would be really confused and would get an impression that something is really broken there.
+
+On the other hand FederatedReplicaSets should somehow reflect the status of the underlying cluster - like number of total replicas that are there so that kubectl rollout has proper data to work on. 
+
+<img src="federated-deployment-3.png" width="500"/>
+
+Federated nginx-701339712 gets all of the statistics as well as spec elements (like the total number of replicas) from ReplicaSets in the underlying clusters. Obviously none of the spec changes should be pushed back. With this change the whole thing would look like as a well-working “whole” while in fact Federated ReplicaSets would be “shadows” (or puppets) - their original functionality (creating rs in clusters and controlling the balancing) would be turned off. 
```

Comment:
How about deployment controller updating `Federated nginx-701339712` based on the federated deployment (total replicas and updated replicas) instead of federated replicaset controller updating it based on underlying replicasets? That way we wont be relying on the internal implementation of underlying deployments (whether they create a rs or not and with what name).